include/uart.h

```diff
@@ -29,6 +29,157 @@ extern "C" {
+	 */
+	UART_TX_TIMEOUT,
+	/**
+	 * @brief Received data is ready for processing.
+	 *
+	 * This event is generated in two cases:
+	 * - When RX timeout occurred, and data was stored in provided buffer.
+	 * - When provided buffer is full.
+	 */
+	UART_RX_RDY,
+	/**
+	 * @brief Driver requests next buffer for continuous transmission.
+	 *
+	 * This event is triggered when receiving has started for a new buffer.
+	 * For continuous reliable receiving, user should provide another RX
+	 * buffer in response to this event, using uart_rx_buf_rsp() function.
+	 *
+	 * If uart_rx_buf_rsp() function is not called, receiving will stop
+	 * after filling current buffer and UART_RX_DONE will be generated.
+	 */
+	UART_RX_BUF_REQUEST,
+	/**
+	 * @brief Receiving has stopped and provided buffer is full.
+	 *
+	 * If no second buffer was provided, receiving will stop after filling
+	 * current buffer. After that event, receiving can be enabled again
+	 * using rx_en() function.
+	 */
+	UART_RX_DONE,
+	/**
+	 * @brief Buffer is no longer used by UART driver.
+	 *
+	 * This event occurs after rx_disable() function is called, for every
+	 * buffer passed to UART driver, it indicates that buffer can be freed
+	 * and is no longer used.
+	 */
+	UART_RX_BUF_RELEASED,
+	/**
+	 * @brief RX has been disabled and can be reenabled.
+	 *
+	 * After rx_disable() function is called, driver may need some time to
+	 * stop the transmission, after releasing buffers this event is
+	 * generated and receiving can be enabled again using rx_en() function.
+	 */
+	UART_RX_DISABLED,
+	/**
+	 * @brief Break interrupt
+	 *
+	 * A break interrupt was received. This happens when the serial input
+	 * is held at a logic '0' state for longer than the sum of
+	 * start time + data bits + parity + stop bits.
+	 */
+	UART_BREAK,
+	/** @brief Overrun error */
+	UART_ERROR_OVERRUN,
+	/** @brief Parity error */
+	UART_ERROR_PARITY,
+	/** @brief Framing error */
+	UART_ERROR_FRAMING,
+};
+
+/* @brief Backward compatibility defines, deprecated */
+#define UART_ERROR_BREAK UART_BREAK
+#define LINE_CTRL_BAUD_RATE (1 << 0)
+#define LINE_CTRL_RTS UART_LINE_CTRL_RTS
+#define LINE_CTRL_DTR UART_LINE_CTRL_DTR
+#define LINE_CTRL_DCD UART_LINE_CTRL_DCD
+#define LINE_CTRL_DSR UART_LINE_CTRL_DSR
+
+/* @brief Structure containing information about current event. */
+struct uart_event {
+	/* @brief Type of event */
+	enum uart_event_type type;
+	/* @brief Pointer to current buffer. */
+	u8_t *buf;
+	/* @brief Offset from buffer start to currently transferred data. */
+	size_t offset;
+	/* @brief Number of bytes transferred. */
+	size_t len;
+};
+
+/**
+ * @typedef uart_callback_t
+ * @brief Define the application callback function signature for
+ * uart_set_callback() function.
+ *
+ * @param dev	    UART device structure.
+ * @param evt	    Pointer to uart_event structure.
+ * @param user_data Pointer to data specified by user.
+ */
+typedef void (*uart_callback_t)(struct device *dev, struct uart_event *evt,
```

Comment:
we had a discussion in counter API that kind of concluded that `struct device *dev` is redundant in the callback. `user_data` can point to `dev` in single-object case and if driver user is mulitinstance module then `user_data` points to the object which must have reference to the device.

Reply:
Ok, I will remove pointer to device

Reply:
Wait, what about consistency with other existing APIs?